agent/php_api.c

```diff
@@ -1578,3 +1578,70 @@ PHP_FUNCTION(newrelic_get_trace_metadata) {
     nr_free(span_id);
   }
 }
+
+/*
+ * Purpose      (New Relic API) (Error Fingerprinting)
+ *              Allows a user to register a PHP callback to be invoked by the
+ *              Agent when the application encounters an error for the purpose
+ *              of returning a string for the 'error.group.name' Agent
+ *              Attribute.
+ *
+ */
+#ifdef TAGS
+void zif_newrelic_set_error_group_callback(void); /* ctags landing pad only */
+void newrelic_set_error_group_callback(void);     /* ctags landing pad only */
+#endif
+PHP_FUNCTION(newrelic_set_error_group_callback) {
+  zend_fcall_info fci;
+  zend_fcall_info_cache fcc;
+
+  NR_UNUSED_RETURN_VALUE_PTR;
+  NR_UNUSED_THIS_PTR;
+  NR_UNUSED_RETURN_VALUE_USED;
+
+  nr_php_api_add_supportability_metric("set_error_group_callback" TSRMLS_CC);
+
+  // Verify that only one argument has been provided to the API (the callback)
+  if (1 != ZEND_NUM_ARGS()) {
+    nrl_warning(NRL_API,
+                "newrelic_set_error_group_callback failure: invalid number of "
+                "parameters");
+    RETURN_FALSE;
+  }
+
+  // Verify that the argument passed to the API is a function
+  // Populate function data
+  if (FAILURE == zend_parse_parameters(ZEND_NUM_ARGS(), "f", &fci, &fcc)) {
+    nrl_warning(
+        NRL_API,
+        "newrelic_set_error_group_callback failure: invalid argument passed");
+    RETURN_FALSE;
+  }
+
+  // Verify the user callback accepts 2 arguments
+  if (2 != fcc.function_handler->common.num_args) {
+    nrl_warning(NRL_API,
+                "newrelic_set_error_group_callback failure: invalid number of "
+                "callback parameters: %d",
+                fcc.function_handler->common.num_args);
+    RETURN_FALSE;
+  }
+
+  // Log debug message if the user is overwriting an existing callback
+  if (is_error_callback_set()) {
+    nrl_debug(
+        NRL_API,
+        "newrelic_set_error_group_callback: overwriting previous callback");
+  }
+
+  // Set global values
+  NRPRG(error_group_user_callback).fci = fci;
```

Comment:
I feel like since there is an accessor function for `is_set` here that perhaps there ought to be a setter function for these global values since there was a partial effort to hide the implementation.

Reply:
The `is_set` function wasn't meant to hide the implementation, just to make it easier to check if the values were set. Since the `is_set` function is currently called 4 times (and could conceivably need to be checked again at some point/effort in the future), adding logic to make it easier to check these values made sense to me. 
The actual setting of these values is only ever done once in this function and I don't think that's liable to change, so to me a setter function for this individual case makes less sense.